starknet_py/class_hash.py

```diff
@@ -0,0 +1,124 @@
+    ]
+    builtins_hash = compute_hash_on_elements(_encoded_builtins)
+
+    hinted_class_hash = _compute_hinted_class_hash(contract_class)
+
+    program_data_hash = compute_hash_on_elements(
+        [int(data_, 0) for data_ in contract_class.program["data"]]
+    )
+
+    return compute_hash_on_elements(
+        [
+            api_version,
+            external_entry_points_hash,
+            l1_handler_entry_points_hash,
+            constructor_entry_points_hash,
+            builtins_hash,
+            hinted_class_hash,
+            program_data_hash,
+        ]
+    )
+
+
+def _entry_points_array(entry_points: List[EntryPoint]) -> List[int]:
+    entry_points_array = []
+    for entry_point in entry_points:
+        entry_points_array.extend([entry_point.selector, entry_point.offset])
+
+    return entry_points_array
+
+
+def _compute_hinted_class_hash(contract_class: DeclaredContract) -> int:
+    program = contract_class.program
+    program["debug_info"] = None
+
+    if "attributes" in program:
+        program = _delete_attributes(program)
+
+    # If compiler_version is not present, this was compiled with a compiler before version 0.10.0.
+    # Use "(a : felt)" syntax instead of "(a: felt)" so that the class hash will be the same.
+    if "compiler_version" not in program:
+        program["identifiers"] = _fix_cairo_types(program["identifiers"])
+
+    # ContractClass stores keys of `hints` field as integers, resulting in different sorting order
+    program["hints"] = {int(key): val for key, val in program["hints"].items()}
+
+    class_ = dict(program=program, abi=contract_class.abi)
+    serialized_contract_class = json.dumps(obj=class_, sort_keys=True)
+    return starknet_keccak(data=serialized_contract_class.encode())
+
+
+def _fix_cairo_types(identifiers: dict) -> dict:
+    """
+    Recursively goes through identifiers looking for "cairo_type" fields.
+    Pads values with a space before the colon between variable and type.
+
+    Example:
+        (retdata_size: felt, retdata: felt*) => (retdata_size : felt, retdata : felt*)
+    """
+    for name, value in identifiers.items():
+        if not isinstance(value, dict):
+            continue
+
+        if "cairo_type" in value:
+            value["cairo_type"] = _add_backward_compatibility_space(value["cairo_type"])
+
+        identifiers[name] = _fix_cairo_types(value)
+
+    return identifiers
+
+
+def _add_backward_compatibility_space(cairo_type: str) -> str:
+    result = ""
+    for char in cairo_type:
+        if char == ":" and result[-1] != " ":
+            result += " "
+        result += char
+    return result
+
+
+def _delete_attributes(program) -> dict:
+    if len(program["attributes"]) == 0:
+        # Remove attributes field from raw dictionary, for hash backward compatibility of
+        # contracts deployed prior to adding this feature.
+        del program["attributes"]
```

Comment:
Let's add an early return here instead of `if else`. It'll make this function a bit more readable here I think.